File: typescript/packages/browser/src/vfs/s3/s3.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { BoundVFS } from '@struktoai/mirage-core/vfs/bound'
import { S3_IO } from '@struktoai/mirage-core/commands/builtin/s3/io'
import { S3Accessor } from '@struktoai/mirage-core/accessor/s3'

import { S3_COMMANDS } from '@struktoai/mirage-core/commands/builtin/s3/index'
import type { RegisteredCommand } from '@struktoai/mirage-core/commands/config'

import { buildDeltaHook } from '@struktoai/mirage-core/core/s3/watch'
import type { RegisteredOp } from '@struktoai/mirage-core/ops/registry'
import { S3_OPS } from '@struktoai/mirage-core/ops/s3/index'

import type { FindOptions, VFS } from '@struktoai/mirage-core/vfs/base'
import { s3StorageId } from '@struktoai/mirage-core/vfs/s3/storage_id'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { VFSName } from '@struktoai/mirage-core/types'

import { type DeltaHook } from '@struktoai/mirage-core/watch/index'
import { redactConfig, type S3Config, type S3ConfigRedacted } from './config.ts'

export const S3_BROWSER_PROMPT = `{prefix}
  Remote S3 bucket accessed via presigned URLs (browser runtime).
  Supports the full filesystem command set: ls/tree/cat/grep/find/du/cp/mv/rm/etc.
  Listing operations require the presigner to sign LIST/COPY operations in
  addition to GET/PUT/HEAD/DELETE — see S3BrowserPresignedUrlProvider docs.`

export interface S3VFSState {
  type: string
  config: S3ConfigRedacted
}

export class S3VFS extends BoundVFS<S3Accessor> implements VFS {
  declare appendFile: (path: PathSpec, data: Uint8Array) => Promise<void>

  declare writeFile: (p: PathSpec, data: Uint8Array) => Promise<void>

  declare exists: (p: PathSpec) => Promise<boolean>

  declare mkdir: (p: PathSpec) => Promise<void>

  declare rmdir: (p: PathSpec) => Promise<void>

  declare unlink: (p: PathSpec) => Promise<void>

  declare rename: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare truncate: (p: PathSpec, length: number) => Promise<void>

  declare copy: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare rmR: (p: PathSpec) => Promise<void>

  declare du: (p: PathSpec) => Promise<number>

  declare find: (p: PathSpec, options?: FindOptions) => Promise<string[]>

  readonly supportsSnapshot: boolean = true
  readonly kind: string = VFSName.S3
  readonly cachesReads: boolean = true
  // A HEAD carries ContentLength, so a size is always knowable without
  // fetching. Every sibling browser VFS says so; s3 was the one that
  // did not, and its node twin has always declared it.
  readonly sizesAlwaysKnown: boolean = true
  // stat and read both stamp the ETag, so the gate compares like with
  // like. Inherited by every S3AliasVFS provider.
  readonly readRevalidatable: boolean = true
  readonly prompt: string = S3_BROWSER_PROMPT
  readonly config: S3Config
  readonly accessor: S3Accessor

  constructor(config: S3Config) {
    super(S3_IO)
    this.config = config
    this.accessor = new S3Accessor(this.config)
  }

  // Without this, two mounts of one bucket at different prefixes are two
  // storages, so `mv` between them copies an object over itself and then
  // unlinks the source. Node has always declared it; the shared helper keeps
  // the two runtimes from computing different identities for one bucket.
  override storageId(): string {
    return s3StorageId(this.kind, this.config)
  }

  commands(): readonly RegisteredCommand[] {
    return S3_COMMANDS.toArray()
  }

  ops(): readonly RegisteredOp[] {
    return S3_OPS
  }

  deltaHook(): DeltaHook {
    return buildDeltaHook(this.accessor)
  }

  override getState(): Promise<S3VFSState> {
    return Promise.resolve({
      type: this.kind,
      config: redactConfig(this.config),
    })
  }
}

File: python/mirage/vfs/s3/s3.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.s3 import S3Accessor, S3Config
from mirage.commands.builtin.s3 import COMMANDS as S3_COMMANDS
from mirage.commands.builtin.s3.io import IO
from mirage.core.s3.watch import build_delta_hook
from mirage.ops.s3 import OPS as S3_OPS
from mirage.types import VFSName
from mirage.vfs.bound import BoundVFS
from mirage.vfs.s3.prompt import PROMPT
from mirage.watch.base import DeltaHook


class S3VFS(BoundVFS):

    accessor: S3Accessor
    name: str = VFSName.S3
    # byte store: stat() sizes every file from metadata
    SIZES_ALWAYS_KNOWN: bool = True
    caches_reads: bool = True
    PROMPT: str = PROMPT
    SUPPORTS_SNAPSHOT: bool = True
    # stat and read both stamp the ETag, so the gate compares like with
    # like. Inherited by every S3AliasVFS provider.
    READ_REVALIDATABLE: bool = True

    def __init__(self, config: S3Config) -> None:
        super().__init__(io=IO)
        self.config = config
        self.accessor = S3Accessor(self.config)
        for fn in S3_COMMANDS:
            self.register(fn)
        for op in S3_OPS:
            self.register_op(op)

    def storage_id(self) -> str:
        # Endpoint, bucket and key prefix pin the object namespace. The
        # endpoint matters because the same bucket name on two providers
        # (AWS vs MinIO vs R2) is two different stores. The prefix joins
        # path-like so two mounts whose prefixes nest still resolve to
        # one key once the mount-relative path is appended.
        cfg = self.config
        prefix = (cfg.key_prefix or "").strip("/")
        base = f"{self.name}:{cfg.endpoint_url or 'aws'}:{cfg.bucket}"
        return f"{base}/{prefix}" if prefix else base

    def delta_hook(self) -> DeltaHook:
        return build_delta_hook(self.accessor)

    def get_state(self) -> dict[str, Any]:
        return self.config_state(self.config)
